### PR TITLE
ENH: Make layout order an initialization parameter of ArrayProxy

### PR DESCRIPTION
**Pull request by effigies**
For #1090, it's been useful to set the proxy order dynamically, rather than subclassing `ArrayProxy`. This makes it an initialization argument. I've left it as a class variable on the off-chance that somebody actually uses something like:

```Python
class CArrayProxy(ArrayProxy):
    order = 'C'
```

Minor in-passing cleanups, such as removing the explicit `(object)` superclass and moving the test loops to parametrizations.

### CODE REVIEW

**Comment by effigies** [comment]
Anybody up for a review?

**Comment by matthew-brett** [comment]
Will do, tomorrow ...

File: nibabel/arrayproxy.py [hunk 1]
```diff
@@ -116,6 +115,10 @@ def __init__(self, file_like, spec, *, mmap=True, keep_file_open=None):
             True gives the same behavior as ``mmap='c'``.  If `file_like`
             cannot be memory-mapped, ignore `mmap` value and read array from
             file.
+        order : {'F', 'C'}, optional, keyword only
```
Comment (matthew-brett) [inline nibabel/arrayproxy.py:118]: Shouldn't this be:

```suggestion
        order : {None, 'F', 'C'}, optional, keyword only
```

?

File: nibabel/tests/test_arrayproxy.py [hunk 2]
```diff
@@ -154,33 +154,55 @@ def test_nifti1_init():
...
     fobj = BytesIO()
-    fobj.write(b'\0' * offset)
+    fobj.write(bytes(offset))
     fobj.write(arr.tobytes(order='F'))
     prox = ArrayProxy(fobj, hdr)
     sliceobj = (None, slice(None), 1, -1)
     assert_array_equal(arr[sliceobj] * 2.0 + 1.0, prox[sliceobj])
 
 
+@pytest.mark.parametrize("order", ("C", "F"))
```
Comment (matthew-brett) [inline nibabel/tests/test_arrayproxy.py:193]: Also test `order=None`?

**Review by matthew-brett — COMMENTED** [review]
Hmm - I see what you mean, but the interface is now a bit clunky, because there are two ways to override the order.   I guess the most direct interface would be to make the class variable be `default_order`.     Is there a good way of going there without breaking people's code?

File: nibabel/tests/test_arrayproxy.py (hunk 2 above)
Comment (effigies) [reply nibabel/tests/test_arrayproxy.py:193]: Does not work for `fobj.write(arr.tobytes(order=order))` (currently L198). This is testing override specifically, so I'm not sure "use default" is really relevant.

**Comment by effigies** [comment]
> Hmm - I see what you mean, but the interface is now a bit clunky, because there are two ways to override the order. I guess the most direct interface would be to make the class variable be `default_order`. Is there a good way of going there without breaking people's code?

Could change `order` to `default_order`, and then look for an `order` class variable (`hasattr(self.__class__, "order")`) in `__init__` and raise a `DeprecationWarning`.

**Comment by matthew-brett** [comment]
> > 
> 
> Could change `order` to `default_order`, and then look for an `order` class variable (`hasattr(self.__class__, "order")`) in `__init__` and raise a `DeprecationWarning`.

Nice idea.

**Comment by effigies** [comment]
Matthew, any further comments?

File: nibabel/tests/test_arrayproxy.py [hunk 3]
```diff
@@ -57,6 +60,10 @@ def copy(self):
 
 class CArrayProxy(ArrayProxy):
     # C array memory layout
+    _default_order = 'C'
+
+
+class DeprecatedCArrayProxy(ArrayProxy):
```
Comment (matthew-brett) [inline nibabel/tests/test_arrayproxy.py:66]: Timed out depraction again (as above)?

**Review by matthew-brett — COMMENTED** [review]
Oops - sorry - some of these I forgot to add in another review.

File: nibabel/arrayproxy.py [hunk 4]
```diff
@@ -142,11 +148,25 @@ def __init__(self, file_like, spec, *, mmap=True, keep_file_open=None):
...
             raise TypeError('spec must be tuple of length 2-5 or header object')
 
+        # Warn downstream users that the class variable order is going away
+        if hasattr(self.__class__, 'order'):
+            warnings.warn(f'Class {self.__class__} has an `order` class variable. '
+                          'ArrayProxy subclasses should rename this variable to `_default_order` '
+                          'to avoid conflict with instance variables.\n'
+                          '* deprecated in version: 5.0\n'
+                          '* will raise error in version: 7.0\n',
+                          DeprecationWarning, stacklevel=2)
```
[the remark this answers is not in the document]
Comment (effigies) [reply nibabel/arrayproxy.py:159]: It's a decorator that works for deprecating an entire function/method/class. It does not work for cases like this. This is why I added the check below:

File: nibabel/tests/test_arrayproxy.py [hunk 5]
```diff
@@ -154,33 +166,87 @@ def test_nifti1_init():
...
+    # We don't really care about the original order, just that the behavior
+    # of the deprecated mode matches the new behavior
+    fprox = ArrayProxy(fobj, (shape, arr.dtype), order='F')
+    cprox = ArrayProxy(fobj, (shape, arr.dtype), order='C')
+
+    # Start raising errors when we crank the dev version
+    if Version(__version__) >= Version('7.0.0.dev0'):
+        cm = pytest.raises(ExpiredDeprecationError)
+    else:
+        cm = pytest.deprecated_call()
```
Comment (effigies) [inline nibabel/tests/test_arrayproxy.py:234]: ... here.

File: nibabel/tests/test_arrayproxy.py (hunk 3 above)
Comment (effigies) [reply nibabel/tests/test_arrayproxy.py:66]: This is specifically for testing. The warning/expiration will show up in `test_deprecated_order_classvar`. When we remove that function, we can remove this class. I'll add a comment...

Comment (effigies) [inline nibabel/tests/test_arrayproxy.py:66]:
```suggestion
class DeprecatedCArrayProxy(ArrayProxy):
    # Used in test_deprecated_order_classvar. Remove when that test is removed (8.0)
```

**Review by matthew-brett — COMMENTED** [review]
LGTM ...

**Comment by effigies** [comment]
Thanks!
